tempesta_fw/http.c

```diff
@@ -2738,21 +2737,25 @@ tfw_http_req_process(TfwConn *conn, const TfwFsmData *data)
 			       && (req->content_length != req->body.len));
 		}
 
-		/* Assign the right Vhost for this request. */
-		if (tfw_http_req_set_context(req)) {
-			TFW_INC_STAT_BH(clnt.msgs_otherr);
-			tfw_client_drop(req, 500, "cannot find"
-				      "Vhost for request");
-			return TFW_BLOCK;
+		/* Assign the right Vhost for current request. */
+		if ((req->vhost = tfw_http_tbl_vhost((TfwMsg *)req, &block))) {
+			req->location = tfw_location_match(req->vhost,
+							   &req->uri_path);
+			r = tfw_gfsm_move(&conn->state, TFW_HTTP_FSM_REQ_MSG,
+					  &data_up);
+			TFW_DBG3("TFW_HTTP_FSM_REQ_MSG return code %d\n", r);
+			/* Don't accept any following requests from the peer. */
+			if (r == TFW_BLOCK) {
+				TFW_INC_STAT_BH(clnt.msgs_filtout);
+				tfw_client_block(req, 403, "parsed request"
+						 " has been filtered out");
+				return TFW_BLOCK;
+			}
 		}
-
-		r = tfw_gfsm_move(&conn->state, TFW_HTTP_FSM_REQ_MSG, &data_up);
-		TFW_DBG3("TFW_HTTP_FSM_REQ_MSG return code %d\n", r);
-		/* Don't accept any following requests from the peer. */
-		if (r == TFW_BLOCK) {
+		if (block) {
 			TFW_INC_STAT_BH(clnt.msgs_filtout);
-			tfw_client_block(req, 403, "parsed request"
-				       " has been filtered out");
+			tfw_client_block(req, 500, "request has been filtered"
```

Comment:
I'd prefer `403 Forbidden` here. It's not server or Tempesta fault that client has requested forbidden location. Think of it from `http_resp_code_block` directive perspective. It worth to block clients that cause many 403 errors, while legitimate users have margin for mistake,

Reply:
Corrected.